### PR TITLE
Save activated ExperimentsDescriptions in DB

### PR DESCRIPTION
**Pull request by qdpham13**
- Save activated ExperimentsDescriptions in Experiments table
- Call updateActivatedExperiments method in experiments activate method
- Load activated ExperimentsDescriptions along with other experiment meta
- bug/263139421

### CODE REVIEW

File: FirebaseRemoteConfig/Tests/Unit/RCNConfigDBManagerTest.m [hunk 1]
```diff
@@ -504,6 +504,45 @@ - (void)testWriteAndLoadExperiments {
...
+  XCTestExpectation *updateAndLoadExperimentExpectation =
+      [self expectationWithDescription:@"Update and load experiment in database successfully"];
+
+  NSError *error;
+  NSArray *payload2 = @[ @"ab", @"cd" ];
+  NSData *payloadData2 = [NSJSONSerialization dataWithJSONObject:payload2
+                                                         options:NSJSONWritingPrettyPrinted
+                                                           error:&error];
+  NSDictionary *payload3 =
+      @{@"experiment_ID" : @"35667", @"experiment_activate_name" : @"activate_game"};
```
Comment (amanda-xia) [inline FirebaseRemoteConfig/Tests/Unit/RCNConfigDBManagerTest.m:517]: New to iOS code - Are the different `payload` variables here supposed to represent different experiments? I don't recognize the structure / what ab" "cd" and "experiment_activate_name" are.

Comment (qdpham13) [reply FirebaseRemoteConfig/Tests/Unit/RCNConfigDBManagerTest.m:517]: Yeah it's to represent different ExperimentDescriptions. This is how other payloads are written too. I don't think the actual JSON contents is too important in this test b/c it just makes sure it can read and write to the DB.

File: FirebaseRemoteConfig/Sources/RCNConfigExperiment.m [hunk 2]
```diff
@@ -32,7 +32,9 @@ @interface RCNConfigExperiment ()
     NSMutableArray<NSData *> *experimentPayloads;  ///< Experiment payloads.
 @property(nonatomic, strong)
     NSMutableDictionary<NSString *, id> *experimentMetadata;  ///< Experiment metadata
-@property(nonatomic, strong) RCNConfigDBManager *DBManager;   ///< Database Manager.
+@property(nonatomic, strong)
+    NSMutableArray<NSData *> *activatedExperimentPayloads;   ///< Activated experiment payloads.
```
Comment (danasilver) [inline FirebaseRemoteConfig/Sources/RCNConfigExperiment.m:36]: Nit: It'd be helpful to use either "active" and "activated" consistently

Comment (danasilver) [reply FirebaseRemoteConfig/Sources/RCNConfigExperiment.m:36]: And does this need to be initialized in `initWithDBManager:` similar to `_experimentPayloads = [[NSMutableArray alloc] init];`?

File: FirebaseRemoteConfig/Sources/RCNConfigExperiment.m [hunk 3]
```diff
@@ -157,6 +179,18 @@ - (void)updateExperimentStartTime {
                          completionHandler:nil];
 }
 
+- (void)updateActiveExperiments {
```
Comment (danasilver) [inline FirebaseRemoteConfig/Sources/RCNConfigExperiment.m:182]: Nit: Might call this `updateActiveExperimentsInDb` or something to make it clear this is not related to updating active experiments in the Analytics SDK (I think those methods have similar names)

**Review by danasilver — COMMENTED** [review]
Trying to grok the structure of experiments in the database...

```
experiments table
---
key: experiment_payload          value: experiment a
key: experiment_payload          value: experiment b
key: experiment_active_payload   value: experiment a
key: experiment_active_payload   value: experiment c
```

Does that match your understanding?

File: FirebaseRemoteConfig/Sources/RCNConfigExperiment.m (hunk 2 above)
Comment (qdpham13) [reply FirebaseRemoteConfig/Sources/RCNConfigExperiment.m]: Switched to only "active". And I think `loadExperimentsFromTable` will always init the experiments variables even if it doesn't have that key in the table. But I added the explicit init to be in line with the other vars

File: FirebaseRemoteConfig/Sources/RCNConfigExperiment.m [hunk 4]
```diff
@@ -131,6 +150,9 @@ - (void)updateExperimentsWithHandler:(void (^)(NSError *_Nullable))handler {
                            lastStartTime:lastStartTime
                                 payloads:_experimentPayloads
                        completionHandler:handler];
+
+  /// Update activated experiments payload and metadata in DB.
+  [self updateActiveExperiments];
```
[the remark this answers is not in the document]
Comment (qdpham13) [reply FirebaseRemoteConfig/Sources/RCNConfigExperiment.m]: Yeah the method that it's in is called with every activated. https://github.com/firebase/firebase-ios-sdk/blob/master/FirebaseRemoteConfig/Sources/FIRRemoteConfig.m#L339

File: FirebaseRemoteConfig/Sources/RCNConfigDBManager.m [hunk 5]
```diff
@@ -827,12 +827,20 @@ - (void)loadExperimentWithCompletionHandler:(RCNDBCompletion)handler {
...
+      activatedExperimentPayloads = [[NSMutableArray alloc] init];
+    }
+
     if (handler) {
       dispatch_async(dispatch_get_main_queue(), ^{
         handler(
             YES, @{
               @RCNExperimentTableKeyPayload : [experimentPayloads copy],
-              @RCNExperimentTableKeyMetadata : [experimentMetadata copy]
+              @RCNExperimentTableKeyMetadata : [experimentMetadata copy],
```
[the remark this answers is not in the document]
Comment (qdpham13) [reply FirebaseRemoteConfig/Sources/RCNConfigDBManager.m:842]: Updated

File: FirebaseRemoteConfig/Sources/RCNConfigExperiment.m (hunk 3 above)
Comment (qdpham13) [reply FirebaseRemoteConfig/Sources/RCNConfigExperiment.m]: Updated

**Comment by qdpham13** [comment]
> Trying to grok the structure of experiments in the database...
> 
> ```
> experiments table
> ---
> key: experiment_payload          value: experiment a
> key: experiment_payload          value: experiment b
> key: experiment_active_payload   value: experiment a
> key: experiment_active_payload   value: experiment c
> ```
> 
> Does that match your understanding?

Yeah, that's how it works in iOS. One table and keys that differentiate the data

File: FirebaseRemoteConfig/Sources/RCNConfigExperiment.m [hunk 6]
```diff
@@ -131,6 +151,9 @@ - (void)updateExperimentsWithHandler:(void (^)(NSError *_Nullable))handler {
                            lastStartTime:lastStartTime
                                 payloads:_experimentPayloads
```
Comment (karenyz) [inline FirebaseRemoteConfig/Sources/RCNConfigExperiment.m:152]: so alternatively we could call `updateActiveExperimentsInDB` before this block, then pass `_activeExperimentPayloads` here? 

do you think that would be better? right now `_activeExperimentPayloads` isn't used anywhere iiuc (though i guess we will use it once we add diff-ing for realtime

Comment (qdpham13) [reply FirebaseRemoteConfig/Sources/RCNConfigExperiment.m:152]: We could do that but I think it's fine the way it is. We plan on adding diff-ing for real-time right after this anyways and real-time will be the main user of `_activeExperimentPayloads`

Comment (karenyz) [reply FirebaseRemoteConfig/Sources/RCNConfigExperiment.m:152]: ok sg

**Review by karenyz — APPROVED** [review]
